mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp

```diff
@@ -442,14 +440,52 @@ struct BlockwiseGemmV2RewritePattern
         sourceOffset = kb.create<MulIOp>(
             loc, sourceOffset, kb.create<ConstantIndexOp>(loc, KPack));
 
-      Value destOffset = k_i;
-
-      // Emit load/store
+      Type loadType = vectorTypeOrSelf(dataType, KPack);
+      Value value =
+          kb.create<InBoundsLoadOp>(loc, loadType, ldsOrig, sourceOffset);
       auto bufferType = regDest.getType().cast<MemRefType>();
       Type bufferElementType = bufferType.getElementType();
-      Value value = kb.create<InBoundsLoadOp>(loc, bufferElementType, ldsOrig,
-                                              sourceOffset);
-      kb.create<memref::StoreOp>(loc, value, regDest, ValueRange{destOffset});
+
+      // We're loading in units of kPack, but storing in units of k_base.
+      if (KPack == k_base) {
+        Value destOffset = k_i;
+        kb.create<memref::StoreOp>(loc, value, regDest, ValueRange{destOffset});
+      } else if (KPack > k_base) {
+        int64_t numStores = KPack / k_base;
+        Value baseDestOffset = kb.createOrFold<arith::MulIOp>(
+            loc, k_i, kb.createOrFold<arith::ConstantIndexOp>(loc, numStores));
+        for (int64_t i = 0; i < numStores; ++i) {
+          Value sliceStart =
+              kb.createOrFold<arith::ConstantIndexOp>(loc, k_base * i);
+          Value slice = kb.create<ExtractSliceOp>(loc, bufferElementType, value,
+                                                  sliceStart);
+          Value destOffset = kb.createOrFold<arith::AddIOp>(
+              loc, baseDestOffset,
+              kb.createOrFold<arith::ConstantIndexOp>(loc, i));
+          kb.create<memref::StoreOp>(loc, slice, regDest,
+                                     ValueRange{destOffset});
+        }
+      } else if (KPack < k_base) {
```

Comment:
~~Implementation wise looks good though. I don't think this will happen, maybe just add an assert here. (Because we are not packing enough, it takes multiple iterations).~~

Coming to think about it, I think you got a point in having this enabled. Although we'll have to take multiple iterations to fill up a k_base, but bank conflict in loading lds might get better.

Reply:
kPack < k_base is also known as kPack = 1 and a k reduction

Reply:
k reduction case usually has a kPack > k_base > 1. 
 - kPack = 1 < k_base was previous supported
 - (kPack > 1) but < k_base is the case that you have just newly enabled in this PR

Reply:
No, I'll note than none of this code actually depends on kpack != 1

So this retains support for kpack = 1, k_base > 1, but also adds support for the other option, which is still rejected by tuning because there's not really a good reason to have it.

Reply:
Yah no I think we were on the same page and my previous comment was exactly talking about it.

> but also adds support for the other option, which is still rejected by tuning because there's not really a good reason to have it.

Part of my implication here is to re-examine assumption again because it looks to me that by enabling this additional setup where we don't pack enough, there's a chance that it can improve certain cases/data type as well. I think your work might actually improve some int8 performance if we enable some narrower tuning. What that means is that we can experiment to see if we run a case where kPack < k_base, if performance can become better than the kPack > k_base.